### PR TITLE
Null operator

### PR DESCRIPTION
**Pull request by ESeNonFossiIo**
Some optimisation to avoid to waste time in multiplications or sums involving null matrices.

### CODE REVIEW

File: include/deal.II/lac/linear_operator.h [hunk 1]
```diff
@@ -275,6 +288,13 @@ template <typename Range, typename Domain> class LinearOperator
     return *this;
   }
 
+  /**
+  * This bool is used to determines whether a linear operator is a
```
Comment (luca-heltai) [inline include/deal.II/lac/linear_operator.h]: to determines -> to determine

File: include/deal.II/lac/linear_operator.h [hunk 2]
```diff
@@ -377,58 +419,65 @@ operator*(typename Range::value_type  number,
     std::is_convertible<typename Range::value_type, typename Domain::value_type>::value,
     "Range and Domain must have implicitly convertible 'value_type's");
 
-  LinearOperator<Range, Domain> return_op = op;
+  if (op.is_null_operator)
+    {
+      return op;
+    }
+  else
```
Comment (luca-heltai) [inline include/deal.II/lac/linear_operator.h]: Shouldn't we add a check here that if `number==0`, then we return a brand new null_operator, i.e.:

~~~
else if(number == 0) {
 return null_operator(op.reinit_domain_vector, op.reinit_range_vector);
} else { 
...
~~~
and remove all further checking on `number != 0`, down below?

Comment (ESeNonFossiIo) [inline include/deal.II/lac/linear_operator.h]: I agree with you. I have changed the code.

File: include/deal.II/lac/linear_operator.h (hunk 1 above)
Comment (ESeNonFossiIo) [inline include/deal.II/lac/linear_operator.h]: Thanks

**Comment by luca-heltai** [comment]
To me this is ok. I'll let @tamiko and @bangerth comment

**Comment by tamiko** [comment]
> To me this is ok. I'll let @tamiko and @bangerth comment

Looks fine. Feel free to merge.

I'll have a closer look at it next week when I'm back in Heidelberg.

**Pull request merged**